source/common/protobuf/visitor.cc

```diff
@@ -103,15 +132,82 @@ void traverseMessageWorker(ConstProtoVisitor& visitor, const Protobuf::Message&
       if (field->is_repeated()) {
         const int size = reflection->FieldSize(message, field);
         for (int j = 0; j < size; ++j) {
-          traverseMessageWorker(visitor, reflection->GetRepeatedMessage(message, field, j), parents,
-                                false, recurse_into_any);
+          traverseMessageWorker<VISITOR, MESSAGE, MSGGETTER, REPMSGGETTER>(
+              visitor, (*REPMSGGETTER)(reflection, message, field, j), parents, false,
+              recurse_into_any);
+        }
+      } else if (reflection->HasField(message, field)) {
+        traverseMessageWorker<VISITOR, MESSAGE, MSGGETTER, REPMSGGETTER>(
+            visitor, (*MSGGETTER)(reflection, message, field), parents, false, recurse_into_any);
+      }
+    }
+  }
+}
+
+template <typename VISITOR, typename MESSAGE, auto MSGGETTER = &getMessageHelper,
+          auto REPMSGGETTER = &getRepeatedMessageHelper>
+void traverseMessageWorkerExt(VISITOR& visitor, MESSAGE& message,
```

Comment:
Yes, of course can I move it there. The initial plan was to just adapt the `traverseMessageWorker(..)`, but that unfortunately is not feasible.

Just to get you right: `protobuf_utility.h` does not exist yet. You want me to create that file, or do you mean to add to `test/fuzz/utility.h|cpp`?

Reply:
IMHO a new file is better, as it may not be needed for general fuzzers, and avoiding linking the protobuf-traversal code will reduce the binary size (a bit).

Reply:
I have refactored the visitor now. The old one is restored, but because both needed some support routines, I also factored them into a visitor_helper.h|cc in source/common/protobuf. The new visitor's implementation is in test/fuzz/mutable_visitor.h|cc.

Is that what you had in mind? Or does it go at least into the right direction?